### PR TITLE
Add "Quick Search" extension

### PR DESCRIPTION
**Pull request by iwfan**
## Description

Use keyboard shortcut quickly search selected text or clipboard text via Google or Github or DeepL etc.

For example: 

- press <kbd>Option(⌥) + Ctrl(⌃) + O</kbd> open google search in default browser.
- press <kbd>Option(⌥) + Ctrl(⌃) + G</kbd> open GitHub search in default browser.

## Type of change

- New extension

## Screencast

iShot2021-12-11 23 05 12

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)

### CODE REVIEW

**Comment by pernielsentikaer** [comment]
Hi @iwfan 

Is this ready to get reviewed? 😄

**Comment by iwfan** [comment]
> Hi @iwfan
> 
> Is this ready to get reviewed? 😄

@pernielsentikaer yes, it's ready.

**Comment by pernielsentikaer** [comment]
Great 🔥 

I will give it a spin within a few hours, sorry for the long response time, It's 100% my fault - won't happen again 😄

**Comment by iwfan** [comment]
> Great @iwfan 🔥
> 
> I will give it a spin within a few hours, sorry for the long response time, It's 100% my fault - won't happen again 😄

Hi @pernielsentikaer, does there have new information on this extension?

File: extensions/raycast-quick-search/src/github-search.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { showHUD } from "@raycast/api";
+import { GithubSearchOpner } from "./utils/opener";
+import { isNotEmpty, readtext } from "./utils/readtxt";
+
+export default async () => {
+  try {
+    const text = await readtext();
+    if (isNotEmpty(text)) {
+      await GithubSearchOpner(text);
+      showHUD("🎉 Open github search");
```
Comment (pernielsentikaer) [inline extensions/raycast-quick-search/src/github-search.ts:10]:
```suggestion
      showHUD("🎉 Open GitHub search");
```

File: extensions/raycast-quick-search/src/github-repo.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { showHUD } from "@raycast/api";
+import { GithubRepoOpner } from "./utils/opener";
+import { isNotEmpty, readtext } from "./utils/readtxt";
+
+export default async () => {
+  try {
+    const text = await readtext();
+    if (isNotEmpty(text)) {
+      await GithubRepoOpner(text);
+      showHUD("🎉 Open github search");
```
Comment (pernielsentikaer) [inline extensions/raycast-quick-search/src/github-repo.ts:10]:
```suggestion
      showHUD("🎉 Open GitHub search");
```

File: extensions/raycast-quick-search/src/deepl-trans.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { showHUD } from "@raycast/api";
+import { DeeplOpner } from "./utils/opener";
+import { isNotEmpty, readtext } from "./utils/readtxt";
+
+export default async () => {
+  try {
+    const text = await readtext();
+    if (isNotEmpty(text)) {
+      await DeeplOpner(text);
+      showHUD("🎉 Open deepl search");
```
Comment (pernielsentikaer) [inline extensions/raycast-quick-search/src/deepl-trans.ts:10]:
```suggestion
      showHUD("🎉 Open Deepl search");
```

File: extensions/raycast-quick-search/package.json [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    {
+      "name": "github-search",
+      "title": "Github Search",
+      "subtitle": "Search selected text or clipboard text via Github",
+      "description": "Search selected text or clipboard text via Github.",
+      "mode": "no-view"
+    },
+    {
+      "name": "github-repo",
+      "title": "Github Repo",
```
Comment (pernielsentikaer) [inline extensions/raycast-quick-search/package.json:26]:
```suggestion
      "title": "GitHub Repo",
```

File: extensions/raycast-quick-search/package.json [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    {
+      "name": "google-search",
+      "title": "Google Search",
+      "subtitle": "Search selected text or clipboard text via Google",
+      "description": "Search selected text or clipboard text via Google.",
+      "mode": "no-view"
+    },
+    {
+      "name": "github-search",
+      "title": "Github Search",
```
Comment (pernielsentikaer) [inline extensions/raycast-quick-search/package.json:19]:
```suggestion
      "title": "GitHub Search",
```

File: extensions/raycast-quick-search/src/google-search.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import { showHUD } from "@raycast/api";
+import { GoogleSearchOpner } from "./utils/opener";
+import { isNotEmpty, readtext } from "./utils/readtxt";
+
+export default async () => {
+  try {
+    const text = await readtext();
+    if (isNotEmpty(text)) {
+      await GoogleSearchOpner(text);
+      showHUD("🎉 Open google search");
```
Comment (pernielsentikaer) [inline extensions/raycast-quick-search/src/google-search.ts:10]:
```suggestion
      showHUD("🎉 Open Google search");
```

**Review by pernielsentikaer — CHANGES_REQUESTED** [review]
Hi @iwfan 👋 

Yes, I had to discuss it with the team since it's almost what's possible to do with the “Quick Links” extensions except the clipboard part.

A few comments and suggestions:
- You have to find a logo for the extension, right now you're using the default one.
- Rename the extension to contain “clipboard” everywhere
- I'm not sure that you'll ever hit the catch exception in the scripts?

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

**Comment by iwfan** [comment]
Hi @pernielsentikaer, thanks for your patient guidance, I will make the changes as you suggest later.

But for now, I want to say **how is this extension different from Raycast's built-in "Quick Link"?**

`Quick Link` is a very good feature, it is equivalent to a shortcut entrance (a wormhole) through which you can quickly get to any address you want to go to.
`Quick Search` is similar to it in terms of functionality, but it focuses more on using shortcuts to improve search efficiency.

Suppose the following scenario.

Ivan is a front-end development engineer. He is reading a technical article about "React" and he finds a technical term that he doesn't understand.
At this point he wants to google for the technical term, he first selects that term by double-clicking on it and then presses `Command + C`, copying it to the clipboard.

If he wants to search via `Quick Link`, there are several ways to do so.

- Open the Raycast launcher and search for "Quick Link"
- Open the Raycast launcher and search for Alias of "Quick Link"
- Open "Quick Link" directly via Hotkey

But after opening `Quick Link`, Ivan had to enter `Query` (the keyword to be queried).
Fortunately, Ivan has already copied it to the clipboard, so he just needs to press `Command + V`, and then he can press <kbd>Enter</kbd> to open the Google search page.

If Ivan uses `Quick Search`, he can open the Google search page directly via Hotkey when he selected the technical term in that article.

The difference between Quick Search and Quick Link is that Quick Search automatically retrieves the `Query` from the selected text and clipboard, and then opens the specific search address directly.
This skips the step where Ivan enters the `Query` and makes the search much faster.

**Comment by iwfan** [comment]
Hi @pernielsentikaer, I already made changes,  please start a new review.

File: extensions/raycast-quick-search/src/deepl-trans.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  try {
+    const text = await readtext();
+    if (isNotEmpty(text)) {
+      await DeeplOpner(text);
+      showHUD("🎉 Open Deepl search");
+    } else {
+      showHUD("❌ No text found in clipboard");
+    }
+  } catch (error) {
+    showHUD("💩 Sorry, Can not open Deepl for now!");
```
Comment (fe9lix) [inline extensions/raycast-quick-search/src/deepl-trans.ts:15]:
```suggestion
    showHUD("❌ Cannot open Deepl!");
```

Comment (fe9lix) [reply extensions/raycast-quick-search/src/deepl-trans.ts:15]: I'd suggest to keep it brief and maybe you can check whether there's a meaningful short error message that you could extract and append?

Comment (fe9lix) [reply extensions/raycast-quick-search/src/deepl-trans.ts:15]: Same suggestion applies to all other search cases.

File: extensions/raycast-quick-search/src/utils/opener.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+export const GithubSearchOpner = async (text: string) => {
+  await openInBrowser(urlBuilder(`https://github.com/search?q=`, text))
+}
+
+export const GithubRepoOpner = async (text: string) => {
+  await openInBrowser(`https://github.com/${text}`)
+}
+
+export const DeeplOpner = async (text: string) => {
+  await openInBrowser(`https://deepl.com/zh/translator#en/zh/${text}`)
```
Comment (fe9lix) [inline extensions/raycast-quick-search/src/utils/opener.ts:24]: I think it makes sense to make the from/to language a preference and remove the hardcoded "zh"?

File: extensions/raycast-quick-search/README.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
+# raycast-quick-search
+
+Use keyboard shortcut quickly search selected text or clipboard text via Google or Github or DeepL etc.
```
Comment (fe9lix) [inline extensions/raycast-quick-search/README.md:3]:
```suggestion
Use a global keyboard shortcut to quickly search selected text or clipboard text via Google or Github or DeepL etc.
```

File: extensions/raycast-quick-search/package.json [hunk 10]
```diff
@@ -0,0 +1,61 @@
+{
+  "$schema": "https://www.raycast.com/schemas/extension.json",
+  "name": "raycast-quick-search",
```
Comment (fe9lix) [inline extensions/raycast-quick-search/package.json:3]:
```suggestion
  "name": "quick-search",
```

Comment (fe9lix) [reply extensions/raycast-quick-search/package.json:3]: The "raycast" is implicit, no need to add it to the extension name (would be redundant in the resulting store URL).

**Review by fe9lix — CHANGES_REQUESTED** [review]
Hey @iwfan, thanks for the contribution! I've left some inline comments...

**Comment by fe9lix** [comment]
@iwfan Another comment: This is a pretty nice and useful idea and after some internal alignments, we actually intend to add this functionality at some point to the built-in quicklinks feature of Raycast. Meanwhile, we can still publish the extension after the review and keep it in the store if you want.

**Comment by iwfan** [comment]
Hi @fe9lix! Thanks for your review, I already made changes as you suggest. 

BTW, I prefer to add this functionality to the built-in quicklinks feature of Raycast.

**Review by fe9lix — APPROVED** [review]

**Pull request merged**